Shakespeare-data_cleaning/Data cleaning.py

```diff
@@ -0,0 +1,57 @@
+import re
+
+def process_paragraphs(input_file, output_paragraphs, output_bin):
+    with open(input_file, 'r', encoding='utf-8') as infile, \
+            open(output_paragraphs, 'w', encoding='utf-8') as out_paragraphs, \
+            open(output_bin, 'w', encoding='utf-8') as out_bin:
+
+        full_paragraph_text = ""  # Stores the entire paragraph content
+        char_id = None  # Identifier for CharID
+
+        for line in infile:
+            line = line.strip()
+
+            # Skip empty lines
+            if not line:
+                continue
+
+            # If we encounter a new structured paragraph with `#`, process the previous paragraph
+            if "#" in line:
+                # Process the previous paragraph, if any exists
+                if full_paragraph_text:
+                    # Replace '[p]' with a space
+                    full_paragraph_text = re.sub(r'\[p\]', ' ', full_paragraph_text)
```

Comment:
`full_paragraph_text.replace("[p]", " ", full..`
is more efficient and readable here